sycl/include/CL/sycl/exception.hpp

```diff
@@ -12,40 +12,48 @@
 
 #include <CL/sycl/detail/common.hpp>
 #include <CL/sycl/stl.hpp>
-#include <exception>
+
+#include <cstddef>
 
 namespace cl {
 namespace sycl {
 
+// Forward declaration
 class context;
 
 struct exception {
+
+  using ContextPtr = shared_ptr_class<context>;
+
   exception() = default;
 
-  const char *what() const noexcept { return msg.c_str(); }
+  exception(const char *Msg, const cl_int CLErr = CL_SUCCESS,
+            ContextPtr Context = nullptr)
+      : Msg(string_class(Msg) + " " +
+            ((CLErr == CL_SUCCESS)
+                 ? string_class(std::to_string(CL_SUCCESS) + " (CL_SUCCESS)")
+                 : OCL_CODE_TO_STR(CLErr))),
+        CLErr(CLErr), Context(Context) {}
+
+  exception(const string_class &Msg, const cl_int CLErr = CL_SUCCESS,
+            ContextPtr Context = nullptr)
+      : exception(Msg.c_str(), CLErr, Context) {}
```

Comment:
You are getting c-string here to pass it to the next constructor which will again create `string_class` from it.
I think it would be better to "swap" constructors:

```c++
exception(const string_class &Msg, ...) : // some code {}
exception(const chat *Msg, ...) : exception(string_class(Msg))
```

Reply:
Done.